pyproject.toml

```diff
@@ -133,8 +134,8 @@ dev = [
 ]
 # Algorithms
 paga = ["igraph"]
-louvain = ["igraph", "louvain>=0.6,!=0.6.2"]  # Louvain community detection
-leiden = ["igraph>=0.10", "leidenalg>=0.9"]  # Leiden community detection
+louvain = ["igraph", "louvain>=0.6.0,!=0.6.2"]  # Louvain community detection
+leiden = ["igraph>=0.10", "leidenalg>=0.9.0"]  # Leiden community detection
```

Comment:
why this change? you do e.g `numba>=0.56` too

Reply:
Basically because the `min-deps.py` script is hacky, and is appending `.*` to the end of the version it finds here. I am aiming for testing against the latest bugfix release within a release series (at least for projects that do semantic versioning).

Reply:
That doesn’t sound good, I don’t think it should break if we specify `>=1.2.4` somewhere because that patch release has some patch we rely on.

Reply:
It won't break, it will try to install `1.2.4.*` which finds `1.2.4`

Reply:
Yeah, but it's pretty close, hence the hackyness. Open to alternatives.

Notably the actual requirement understood by pip is the same either way this is written.

Reply:
Other approaches I was thinking about ran into the problem that we have deps that don't follow semver, like dask

Reply:
> Notably the actual requirement understood by pip is the same either way this is written.

Either way? `>=1.2.4, <1.3` is not equivalent to `==1.2.4.*`/`==1.2.4`.

> Other approaches I was thinking about ran into the problem that we have deps that don't follow semver, like dask

But what you’re doing does rely on semver (or at least the “third component is patch version” part of it)

So what semantics *are* you gunning for?

Reply:
> Either way?

I was referring the to change in your suggestion here. `louvain>=0.6.0` is practically `louvain>=0.6`

> But what you’re doing does rely on semver (or at least the “third component is patch version” part of it)

I'm not sure it does rely on semver. Like, the how I think we should treat semver relies on semver. But the script itself works fine for non-semver dependencies, like calendar data ones for example: `>=2023.07.22` -> `==2023.07.22.*`.

> So what semantics are you gunning for?

I am flexible on that. But I think in general I would say "minimum supported version" refers to feature release, not bugfix releases. Similar to [SPEC 0](https://scientific-python.org/specs/spec-0000/)

I think this just works well enough and catches a lot of bugs so is useful. Definitely has parts that could be better.

Ideally there would be some tool where someone figured out some ideal semantics for this case, but apparently that's a hard problem https://github.com/pypa/pip/issues/8085.

Reply:
> I am flexible on that. But I think in general I would say "minimum supported version" refers to feature release, not bugfix releases.

Of course, but we also need a way to specify a bugfix we need. Not doing that opens the door to the resolver finding an older bugfix version that doesn’t have a patch we need.

That makes things harder of course, since we can’t encode “this is semver” into the version specification like you currently do.

Reply:
> But I don’t like losing the ability to say foo>=3.6.4

I mean, we can still do that with the existing code. It just means the CI tests against that bugfix release specifically and not whatever is the latest from pypi.

Reply:
I am not sure how we can both pin a minimum bugfix release and test against the latest in a release series from pypi. If all our dependencies followed semver, then this would be easy since we could just strip the bugfix release number from the specification when making the environment.

Reply:
That means that we can’t do this unless we annotate for every project if it’s semver.

Reply:
> That means that we can’t do this unless we annotate for every project if it’s semver.

Yes

------

Do you have a specific change here that you would like to see?

Reply:
see below, this is too fundamental for a thread.